keyvi/src/cpp/dictionary/dictionary_merger.h

```diff
@@ -87,105 +91,124 @@ final {
+
+        fsa::automata_t fsa;
+
+        if (appendMerge_) {
+            fsa.reset(new fsa::Automata(filename, loading_strategy_types::lazy, false));
+        } else {
+            fsa.reset(new fsa::Automata(filename));
+            if (fsa->GetValueStore()->GetValueStoreType() != ValueStoreT::GetValueStoreType()) {
+                throw std::invalid_argument("Dictionaries must have the same type.");
+            }
+        }
 
-    if (fsa->GetValueStore()->GetValueStoreType() != ValueStoreT::GetValueStoreType()) {
-      throw std::invalid_argument("Dictionaries must have the same type.");
+        dicts_to_merge_.push_back(fsa);
     }
 
-    dicts_to_merge_.push_back(fsa);
-  }
-
-  /**
-   * Set a custom manifest to be embedded into the index file.
-   *
-   * @param manifest as JSON string
-   */
-  void SetManifestFromString(const std::string& manifest){
-    manifest_ = manifest;
-  }
-
-  void Merge(const std::string& filename){
-    std::priority_queue<SegmentIterator> pqueue;
-
-    int i = 0;
-    for (auto fsa: dicts_to_merge_) {
-      fsa::EntryIterator e_it(fsa);
-      pqueue.push(SegmentIterator(e_it, i++));
+    /**
+     * Set a custom manifest to be embedded into the index file.
+     *
+     * @param manifest as JSON string
+     */
+    void SetManifestFromString(const std::string& manifest) {
+        manifest_ = manifest;
     }
 
-    ValueStoreT* value_store = new ValueStoreT(params_);
+    void Merge(const std::string& filename) {
+        std::priority_queue<SegmentIterator> pqueue;
 
-    fsa::Generator<PersistenceT, ValueStoreT> generator(memory_limit_, params_, value_store);
+        int i = 0;
+        for (auto fsa: dicts_to_merge_) {
+            fsa::EntryIterator e_it(fsa);
+            pqueue.push(SegmentIterator(e_it, i++));
+        }
 
-    std::string top_key;
+        ValueStoreT* value_store = appendMerge_ ? new ValueStoreT(inputFiles_) : new ValueStoreT(params_);
 
-    while(!pqueue.empty()){
-      auto segment_it = pqueue.top();
-      pqueue.pop();
+        fsa::Generator<PersistenceT, ValueStoreT> generator(memory_limit_, params_, value_store);
 
-      top_key = segment_it.entryIterator().GetKey();
+        std::string top_key;
 
-      // check for same keys and merge only the most recent one
-      while (!pqueue.empty() and pqueue.top().entryIterator().operator==(top_key)) {
+        while (!pqueue.empty()) {
+            auto segment_it = pqueue.top();
+            pqueue.pop();
 
-        auto to_inc = pqueue.top();
-        TRACE("removing element with prio %d (in favor of %d)", to_inc.priority, entry_it.priority);
+            top_key = segment_it.entryIterator().GetKey();
 
-        pqueue.pop();
-        if (++to_inc) {
-          TRACE("push iterator");
-          pqueue.push(to_inc);
-        }
-      }
+            // check for same keys and merge only the most recent one
+            while (!pqueue.empty() and pqueue.top().entryIterator().operator==(top_key)) {
```

Comment:
Sure! 
Here `Strategy` design pattern fits perfectly, just as we discussed I took the shorter path with less changes, and this will be cleaned out with ValueStoreMerger implementation.